Create rule S6885 (#3571)
SonarSource/rspec: rules/S6885/java/rule.adoc

> Use built-in \"Math.clamp\" methods

Reviewer: This would appear as a duplicate to the title, I think we can remove it.
```suggestion
```

Reviewer 2: :+1:

== Why is this an issue?

> In Java 21 a method to clamp value has been added to the Math package. It is now the preferred way to restrict a value to a given interval. Using the built-in method is more readable and less error-prone.

Reviewer: I would split the lines and reword it a little bit
```suggestion
In Java 21 the `java.lang.Math` class was updated with the static method `Math.clamp`, to clamp a numerical value between two max and min values.
Using this built-in method is now the preferred way to restrict a value to a given interval, as it is more readable and less error-prone.
```

Reviewer 2: :+1:

== How to fix it

> Replace with \"Math.clamp\".

Reviewer: Here we can be a little bit more specific
```suggestion
Replace combinations of ternary operators and `Math.min`/`Math.max` with `Math.clamp`.
```

Reviewer 2: :+1:

=== Code examples

==== Noncompliant code example

[source,java,diff-id=1,diff-type=noncompliant]
----

> int clampedValue = value > max ? max : value < min ? min : value; // Noncompliant; Replace with \"Math.clamp\"

Reviewer: We can avoid escaping quotes here
```suggestion
int clampedValue = value > max ? max : value < min ? min : value; // Noncompliant; Replace with "Math.clamp"
```

Reviewer 2: :+1:

----

[source,java,diff-id=2,diff-type=noncompliant]
----

> int clampedValue = Math.max(min, Math.min(max, value)); // Noncompliant; Replace with \"Math.clamp\"

Reviewer: We can avoid escaping quotes here
```suggestion
int clampedValue = Math.max(min, Math.min(max, value)); // Noncompliant; Replace with "Math.clamp"
```

Reviewer 2: :+1:

----

[source,java,diff-id=3,diff-type=noncompliant]
----

> int clampedValue = value; // Noncompliant; Replace with \"Math.clamp\"

Reviewer: We can avoid escaping quotes here
```suggestion
int clampedValue = value; // Noncompliant; Replace with "Math.clamp"
```
